### PR TITLE
provision wal_autocheckpoint to prevent the log file growing to large size

### PR DESCRIPTION
**Pull request by abstractmachines**
This fix is ported from: 876e5e39e0f431b53d6924695dbdce927552bfd5 - PR 829 2.28

### CODE REVIEW

**Comment by abstractmachines** [comment]
# Testing
> This PR introduces automatic WAL file checkpointing and truncation, which already existed (only) in IndexedDB through WAL hook that truncates the WAL with SQLITE_CHECKPOINT_TRUNCATE after the default threshold of 1000 WAL pages, [here](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp#L991)

- Since automatic WAL file truncation now happens with this PR, simple use of the rpi3 will demonstrate a WAL file that does not grow indefinitely after merging this PR, and checkpointing now occurs with truncation as in 2.28.
- You can also test by using `WPE_WAL_AUTOCHECKPOINT` to set threshold to a smaller number, launch cog browser, and increase, for example, localStorage while inspecting the `-wal` file growth, e.g. `root/.local/share/localstorage -wal` file(s) per-origin;
- You can also use Remote Web Inspector and localStorage API, or browse localStorage-heavy site; env var set to a smaller value will result in `-wal` file on rpi3 being reset at more frequent intervals, and set to 0 bytes as expected with checkpoint TRUNCATE mode.

File: Source/WebCore/platform/sql/SQLiteDatabase.cpp [hunk 1]
```diff
@@ -171,7 +171,16 @@ static int walAutomaticTruncationHook(void* context, sqlite3* db, const char* db
...
+    // prevent -wal file from growing indefinitely w automatic checkpointing:
+    // set threshold for number of pages after which checkpoint should be run
+    static constexpr int checkpointThresholdDefault = 1000; // matches SQLITE_DEFAULT_WAL_AUTOCHECKPOINT
+    int checkpointThreshold = checkpointThresholdDefault;
+
+    // control threshold w/ to checkpoint more frequently if desired
+    auto walAutoCheckpointEnv(getenv("WPE_WAL_AUTOCHECKPOINT"));
+    if (walAutoCheckpointEnv) {
+        checkpointThreshold = atoi(walAutoCheckpointEnv);
+    }
```
Comment (magomez) [inline Source/WebCore/platform/sql/SQLiteDatabase.cpp:185]: Couple of things here:
- you're reading the env var every time this function is called, something that's needed only once.
- you're not checking any error on atoi. It the value is not properly converted, we don't know what's returned (can be 0 or who knows).

I see something like this:
```
static constexpr int checkpointThresholdDefault = 1000;
static int checkpointThreshold = 0;

if (!checkpointThreshold) {
  auto envValue = String::fromLatin1(getenv("WPE_WAL_AUTOCHECKPOINT"));
  checkpointThreshold = parseInteger<int>(envValue).value_or(checkpointThresholdDefault);
}
```
which ensures that the env var is only read once the first time the function is called. If there's a valid env value, it's used, and if there's no value or it's invalid, then the default value is taken.

**Review by magomez — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/sql/SQLiteDatabase.cpp (hunk 1 above)
Comment (magomez) [reply Source/WebCore/platform/sql/SQLiteDatabase.cpp:185]: It would also be good to check that the content of the env var is not a negative value.

Comment (abstractmachines) [reply Source/WebCore/platform/sql/SQLiteDatabase.cpp:185]: Good call, I wasn't a fan of `atoi` either for same reasons; all changes made and tested fairly extensively. Thanks for the feedback!

**Review by magomez — APPROVED** [review]

**Pull request merged**